extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -0,0 +1,376 @@
+		this._treeChangeHandler = treeChangeHandler;
+	}
+
+	public get treeChangeHandler(): IControllerTreeChangeHandler {
+		return this._treeChangeHandler;
+	}
+}
+
+export type ControllerRootNodeArg = {
+	treeChangeHandler: IControllerTreeChangeHandler;
+};
+
+export class ControllerRootNode extends ControllerTreeNode {
+
+	constructor(arg?: ControllerRootNodeArg) {
+		super(Object.assign({
+			label: 'root',
+			nodeType: BdcItemType.controllerRoot,
+		}, arg));
+	}
+
+	public async getChildren(): Promise<ControllerNode[]> {
+		return this.children as ControllerNode[];
+	}
+
+	public addControllerNode(url: string, username: string, password: string, rememberPassword: boolean, masterInstance?: IEndPoint): void {
+		let controllerNode = this.getExistingControllerNode(url, username);
+		if (controllerNode) {
+			controllerNode.password = password;
+			controllerNode.rememberPassword = rememberPassword;
+			controllerNode.clearChildren();
+		} else {
+			controllerNode = new ControllerNode({ url, username, password, rememberPassword, parent: this, treeChangeHandler: this.treeChangeHandler });
+			this.addChild(controllerNode);
+		}
+
+		if (masterInstance) {
+			controllerNode.addSqlMasterNode(masterInstance.endpoint, masterInstance.description);
+		}
+	}
+
+	public deleteControllerNode(url: string, username: string): ControllerNode {
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		let index = nodes.findIndex(e => e.url === url && e.username === username);
+		let deleted = undefined;
+		if (index >= 0) {
+			deleted = nodes.splice(index, 1);
+		}
+		return deleted;
+	}
+
+	private getExistingControllerNode(url: string, username: string): ControllerNode {
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		return nodes.find(e => e.url === url && e.username === username);
+	}
+}
+
+export type ControllerNodeArg = {
+	url: string,
+	username: string,
+	password: string,
+	parent: ControllerTreeNode,
+	treeChangeHandler: IControllerTreeChangeHandler,
+	label?: string,
+	description?: string,
+	rememberPassword?: boolean
+};
+
+export class ControllerNode extends ControllerTreeNode {
+	private _url: string;
+	private _username: string;
+	private _password: string;
+	private _rememberPassword: boolean;
+
+	constructor(arg?: ControllerNodeArg) {
+		super(Object.assign({
+			label: undefined,
+			nodeType: BdcItemType.controller,
+			iconPath: IconPath.controllerNode
+		}, arg));
+
+		let address = ControllerNode.toIpAndPort(arg.url);
+		this.label = arg.label || `controller: ${address} (${arg.username})`;
+		this.description = arg.description || this.label;
+		this._url = arg.url;
+		this._username = arg.username;
+		this._password = arg.password;
+		this._rememberPassword = !!arg.rememberPassword;
+	}
+
+
+	public async getChildren(): Promise<ControllerTreeNode[]> {
+		if (this.children && this.children.length > 0) {
```

Comment:
I don't think we should need to clear & re-query every time. 
- If you've already expanded, why do we clear & refresh? We should have all required metadata right? Maybe I'm misunderstanding why we need to do this

Reply:
synced offline.